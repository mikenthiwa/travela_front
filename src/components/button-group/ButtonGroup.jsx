import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import Button from '../buttons/Buttons';
import './buttonGroup.scss';

class ButtonGroup extends PureComponent {
  filterEntries = (entriesType, statusQuery) => {
    const { fetchRequests, fetchApprovals, url } = this.props;
    const urlSearch = new URLSearchParams(url);
    urlSearch.set('page', 1);
    urlSearch.delete('status');
    urlSearch.delete('type');
    urlSearch.delete('budgetStatus');
    const urlSearchString = `?${urlSearch.toString()}${statusQuery}`;
    if(entriesType === 'requests')
      fetchRequests(urlSearchString);
    else
      fetchApprovals(urlSearchString);
  };

  renderPendingApprovalsButton(){
    const { openApprovalsCount, activeStatus, budgetChecker } = this.props;
    return (
      <Button
        buttonClass={`bg-btn bg-btn--with-badge ${activeStatus === 'open' ?
          'bg-btn--active' : ''}`}
        responsiveText="Open"
        disabled={openApprovalsCount === 0}
        badge={openApprovalsCount}
        showBadge={openApprovalsCount > 0}
        badgeClass={activeStatus === 'open' ?
          'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
        buttonId="open-button"
        onClick={() => this.filterEntries( 'approvals',
          `${budgetChecker ? '&budgetStatus=open' : '&status=open'}`)}
        text="Pending Approvals"
      />
    );
  }


  renderApprovalsButton () {
    const {  pastApprovalsCount, activeStatus, budgetChecker } = this.props;
    return (
      <Fragment>
        <Button
          buttonClass={`bg-btn ${activeStatus === 'all' ? 'bg-btn--active' : ''}`}
          text="All"
          buttonId="all-button"
          onClick={() => this.filterEntries('approvals', '')}
        />
        {this.renderPendingApprovalsButton()}
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${activeStatus === 'past' ? 'bg-btn--active' : ''}`}
          responsiveText="Past"
          buttonId="past-button"
          showBadge={pastApprovalsCount > 0}
          badge={pastApprovalsCount}
          badgeClass={activeStatus === 'past' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
          text="Past Approvals"
          disabled={pastApprovalsCount === 0}
          onClick={() => this.filterEntries(
            'approvals',
            `${ budgetChecker ? '&budgetStatus=past' : '&status=past'}`)}
        />
      </Fragment>
    );
  }

  renderTripModificationButtons = () => {
    const {  cancelledTripsCount, modifiedTripsCount, modificationType } = this.props;
    return (
      <Fragment>
        <Button
          buttonClass={`bg-btn ${modificationType === 'all' ? 'bg-btn--active' : ''}`}
          text="All"
          buttonId="all-button"
          onClick={() => this.filterEntries('approvals', '')}
        />
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${modificationType === 'Cancel Trip' ?
            'bg-btn--active' : ''}`}
          responsiveText="Open"
          disabled={cancelledTripsCount === 0}
          badge={cancelledTripsCount}
          showBadge={cancelledTripsCount > 0}
          badgeClass={modificationType === 'Cancel Trip' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
          buttonId="open-button"
          onClick={() => this.filterEntries( 'approvals','&status=open&type=Cancel Trip')}
          text="Trip Cancellation"
        />
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${modificationType === 'Modify Dates' ? 'bg-btn--active' : ''}`}
          responsiveText="Past"
          buttonId="past-button"
          showBadge={modifiedTripsCount > 0}
          badge={modifiedTripsCount}
          badgeClass={modificationType === 'Modify Dates' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
          text="Date Modifffication"
          disabled={modifiedTripsCount === 0}
          onClick={() => this.filterEntries(
            'approvals', '&status=open&type=Modify Dates')}
        />
      </Fragment>
    );
  }

  renderRequestsButton (openRequestsCount, pastRequestsCount) {
    const { activeStatus } = this.props;
    return (
      <Fragment>
        <Button
          buttonClass={`bg-btn ${activeStatus === 'all' ? 'bg-btn--active' : ''}`}
          text="All"
          buttonId="all-button"
          onClick={() => this.filterEntries('requests', '')}
        />
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${activeStatus === 'open' ?
            'bg-btn--active' : ''}`}
          text="Open Requests"
          responsiveText="Open"
          buttonId="open-button"
          disabled={openRequestsCount === 0}
          badge={openRequestsCount}
          showBadge={openRequestsCount > 0}
          badgeClass={activeStatus === 'open' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge--inactive'}
          onClick={() => this.filterEntries('requests', '&status=open')}
        />
        <Button
          buttonClass={`bg-btn ${activeStatus === 'past' ? 'bg-btn--active' : ''}`}
          disabled={pastRequestsCount === 0}
          text="Past Requests"
          responsiveText="Past"
          buttonId="past-button"
          onClick={() => this.filterEntries('requests', '&status=past')}
        />
      </Fragment>
    );
  }

  renderVerificationButton () {
    const { approvedApprovalsCount, verifiedApprovalsCount, activeStatus } = this.props;
    return (
      <Fragment>
        <Button
          buttonClass={`bg-btn ${activeStatus === 'all' ? 'bg-btn--active' : ''}`}
          text="All"
          buttonId="all-button"
          onClick={() => this.filterEntries('approvals', '')}
        />
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${activeStatus === 'approved' ?
            'bg-btn--active' : ''}`}
          responsiveText="Pending"
          disabled={approvedApprovalsCount === 0}
          badge={approvedApprovalsCount}
          showBadge={approvedApprovalsCount > 0}
          badgeClass={activeStatus === 'approved' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
          buttonId="open-button"
          onClick={() => this.filterEntries('approvals', '&status=approved')}
          text="Pending Verifications"
        />
        <Button
          buttonClass={`bg-btn bg-btn--with-badge ${activeStatus === 'verified' ? 'bg-btn--active' : ''}`}
          responsiveText="Past"
          badge={verifiedApprovalsCount}
          showBadge={verifiedApprovalsCount > 0}
          badgeClass={activeStatus === 'verified' ?
            'bg-btn--with-badge--active' : 'bg-btn--with-badge__approvals--inactive'}
          buttonId="past-button"
          text="Past Verifications"
          disabled={verifiedApprovalsCount === 0}
          onClick={() => this.filterEntries('approvals', '&status=verified')}
        />
      </Fragment>
    );
  }


  render() {
    const { buttonsType, openRequestsCount, pastRequestsCount } = this.props;
    return (
      <div className="button-group">
        { buttonsType === 'approvals' && this.renderApprovalsButton()}
        { buttonsType === 'requests' &&
          this.renderRequestsButton(openRequestsCount, pastRequestsCount) }
        { buttonsType === 'verifications' && this.renderVerificationButton()}
        { buttonsType === 'modifications' && this.renderTripModificationButtons()}
      </div>
    );
  }
}

ButtonGroup.propTypes = {
  openRequestsCount: PropTypes.number,
  pastRequestsCount: PropTypes.number,
  pastApprovalsCount: PropTypes.number,
  openApprovalsCount: PropTypes.number,
  approvedApprovalsCount: PropTypes.number,
  verifiedApprovalsCount: PropTypes.number,
  fetchApprovals: PropTypes.func,
  fetchRequests: PropTypes.func,
  url: PropTypes.string,
  budgetChecker: PropTypes.bool,
  buttonsType: PropTypes.string.isRequired,
  activeStatus: PropTypes.string,
  modifiedTripsCount: PropTypes.number,
  cancelledTripsCount: PropTypes.number,
  modificationType: PropTypes.string
};

ButtonGroup.defaultProps = {
  openRequestsCount: null,
  pastRequestsCount: null,
  modifiedTripsCount: 0,
  cancelledTripsCount: 0,
  pastApprovalsCount: null,
  openApprovalsCount: null,
  approvedApprovalsCount: null,
  verifiedApprovalsCount: null,
  fetchApprovals: null,
  fetchRequests: null,
  budgetChecker: false,
  url: '',
  activeStatus: 'all',
  modificationType: 'all'
};


export default ButtonGroup;
