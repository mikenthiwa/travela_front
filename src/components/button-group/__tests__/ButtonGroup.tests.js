import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import ButtonGroup from '../ButtonGroup';


const props = {
  openRequestsCount: 1,
  pastRequestsCount: 1,
  cancelledTripsCount: 1,
  modifiedTripsCount: 1,
  url: '?page=2',
  fetchRequests: sinon.spy(() => Promise.resolve()),
  fetchApprovals: jest.fn(),
  buttonsType: 'requests',
  activeStatus: 'all',
};

let spy;

describe('<ButtonGroup />', () => {
  it('should render successfully', () => {
    const wrapper = shallow(<ButtonGroup {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the fetchUserRequests method when the `all` button is clicked', () => {
    const wrapper = mount(<ButtonGroup {...props} />);
    spy = sinon.spy(wrapper.instance(), 'filterEntries');
    const allButton = wrapper.find('#all-button');
    allButton.simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('requests', '')).toBe(true);
    spy.resetHistory();
  });

  it('calls the fetchUserRequests method when the `open requests` button is clicked', () => {
    const wrapper = mount(<ButtonGroup {...props} />);
    spy = sinon.spy(wrapper.instance(), 'filterEntries');
    const openButton = wrapper.find('#open-button');
    openButton.simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('requests', '&status=open')).toBe(true);
    spy.resetHistory();
  });

  it('calls the fetchUserRequests method when the `past requests` button is clicked', () => {
    const wrapper = mount(<ButtonGroup {...props} />);
    spy = sinon.spy(wrapper.instance(), 'filterEntries');
    const pastButton = wrapper.find('#past-button');
    pastButton.simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('requests', '&status=past')).toBe(true);
  });

  it('calls the fetchRequests props method with the correct query', () => {
    const wrapper = mount(<ButtonGroup {...{...props, url: '?page=2&limit=3'}} />);
    spy = sinon.spy(wrapper.instance(), 'filterEntries');
    const pastButton = wrapper.find('#past-button');
    const { fetchRequests } = props;
    pastButton.simulate('click');
    expect(spy.calledOnce).toEqual(true);
    expect(spy.calledWith('requests', '&status=past')).toBe(true);
    expect(fetchRequests.called).toEqual(true);
    expect(fetchRequests.calledWith('?page=1&limit=3&status=past'));
  });

  it('sets the correct active button based on the activeStatus props', () => {
    const wrapper = mount(<ButtonGroup {...{...props, activeStatus: 'all' }} />);
    const allButton = wrapper.find('#all-button');
    expect(allButton.hasClass('bg-btn--active')).toEqual(true);
  });

  it('sets the correct active button based on the activeStatus props', () => {
    const wrapper = mount(<ButtonGroup {...{...props, activeStatus: 'open' }} />);
    const openButton = wrapper.find('#open-button');
    expect(openButton.hasClass('bg-btn--active')).toEqual(true);
  });

  it('sets the correct active button based on the activeStatus props', () => {
    const wrapper = mount(<ButtonGroup {...{...props, activeStatus: 'past' }} />);
    const pastButton = wrapper.find('#past-button');
    expect(pastButton.hasClass('bg-btn--active')).toEqual(true);
  });
  describe('Trip modifications', () => {
    const wrapper = mount(<ButtonGroup {...{...props, modificationType: 'all', buttonsType: 'modifications' }} />);

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('should display the trip modification buttons', () => {
      expect(wrapper.find('Button[text="Trip Cancellation"]').length).toEqual(1);
    });

    it('should fetch all the trip modifications', () => {
      wrapper.find('#all-button').simulate('click');
      expect(props.fetchApprovals).toHaveBeenCalledWith('??page=2&page=1');
    });

    it('should filter the cancelled trips', () => {
      wrapper.find('#open-button').simulate('click');
      expect(props.fetchApprovals).toHaveBeenCalledWith('??page=2&page=1&status=open&type=Cancel Trip');
    });

    it('should filter the modified date trips', () => {
      wrapper.find('#past-button').simulate('click');
      expect(props.fetchApprovals).toHaveBeenCalledWith('??page=2&page=1&status=open&type=Modify Dates');
    });
  });
});
