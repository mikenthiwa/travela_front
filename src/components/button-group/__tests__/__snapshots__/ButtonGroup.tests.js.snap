// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup /> should render successfully 1`] = `
<div
  className="button-group"
>
  <Button
    altText=""
    badge={null}
    badgeClass=""
    buttonClass="bg-btn bg-btn--active"
    buttonId="all-button"
    disabled={false}
    imageClass="mdl-Icon"
    imageSrc=""
    onClick={[Function]}
    responsiveText=""
    reverseText={false}
    showBadge={false}
    text="All"
    textClass=""
  />
  <Button
    altText=""
    badge={1}
    badgeClass="bg-btn--with-badge--inactive"
    buttonClass="bg-btn bg-btn--with-badge "
    buttonId="open-button"
    disabled={false}
    imageClass="mdl-Icon"
    imageSrc=""
    onClick={[Function]}
    responsiveText="Open"
    reverseText={false}
    showBadge={true}
    text="Open Requests"
    textClass=""
  />
  <Button
    altText=""
    badge={null}
    badgeClass=""
    buttonClass="bg-btn "
    buttonId="past-button"
    disabled={false}
    imageClass="mdl-Icon"
    imageSrc=""
    onClick={[Function]}
    responsiveText="Past"
    reverseText={false}
    showBadge={false}
    text="Past Requests"
    textClass=""
  />
</div>
`;
